### PR TITLE
Decouple business logic for newly supported regions into ads library

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/6612

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm ad notifications are displayed if ads are enabled
- Confirm ad notifications are not displayed if ads are disabled
- Confirm ads can be disabled and then re-enabled
- Confirm ads work as expected for a fresh install on Android
- Confirm ads work as expected for upgrade paths on Android
- Confirm ads work as expected for a fresh install on Desktop
- Confirm ads work as expected for upgrade paths on Desktop

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by SergeyZhukovsky** [comment]
@tmancey is it still a draft or can be reviewed?

**Comment by tmancey** [comment]
> @tmancey is it still a draft or can be reviewed?

Still in review, as I need to merge https://github.com/brave/brave-core/pull/3770 first so I can rebase this ticket which will remove lots of files as I had to add browser tests in this PR.

File: vendor/bat-native-ads/src/bat/ads/internal/static_values.h [hunk 1]
```diff
@@ -44,48 +44,69 @@ const char kDefaultUserModelLanguage[] = "en";
...
-  { "NL", false },  // Netherlands
-  { "PE", false },  // Peru
-  { "PH", false },  // Philippines
-  { "PL", false },  // Poland
-  { "SE", false },  // Sweden
-  { "SG", false },  // Singapore
-  { "VE", false },  // Venezuela
-  { "ZA", false },  // South Africa
-  { "KY", true  }   // Cayman Islands
+const std::map<int, std::map<std::string, bool>> kSupportedRegionsSchemas = {
```
Comment (SergeyZhukovsky) [inline vendor/bat-native-ads/src/bat/ads/internal/static_values.h]: I would do unsigned int

File: components/brave_ads/common/pref_names.h [hunk 2]
```diff
@@ -23,6 +23,10 @@ extern const char kOnboardingTimestamp[];
 
 extern const char kShouldShowMyFirstAdNotification[];
 
+extern const char kSupportedRegionsLastSchemaVersion[];
+extern const char kSupportedRegionsSchemaVersion[];
+extern const int kSupportedRegionsSchemaVersionNumber;
```
Comment (SergeyZhukovsky) [inline components/brave_ads/common/pref_names.h]: why int? I would do unsigned int

File: components/brave_ads/common/pref_names.h [hunk 3]
```diff
@@ -23,6 +23,10 @@ extern const char kOnboardingTimestamp[];
...
+
 extern const char kVersion[];
 extern const int kCurrentVersionNumber;
```
Comment (SergeyZhukovsky) [inline components/brave_ads/common/pref_names.h]: the same here, I would do unsigned int

File: components/brave_ads/common/pref_names.h (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/common/pref_names.h]: @SergeyZhukovsky `PrefRegistry` does not have an unsigned version of `RegisterIntegerPref`, see https://chromium.googlesource.com/chromium/src.git/+/72.0.3626.80/components/prefs/pref_registry_simple.h?autodive=0%2F%2F%2F

File: components/brave_ads/common/pref_names.h (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/common/pref_names.h]: @SergeyZhukovsky `PrefRegistry` does not have an unsigned version of `RegisterIntegerPref`, see https://chromium.googlesource.com/chromium/src.git/+/72.0.3626.80/components/prefs/pref_registry_simple.h?autodive=0%2F%2F%2F

File: vendor/bat-native-ads/src/bat/ads/internal/static_values.h (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/static_values.h]: @SergeyZhukovsky `PrefRegistry` does not have unsigned versions of `RegisterIntegerPref` and `SetInteger`, see https://chromium.googlesource.com/chromium/src.git/+/72.0.3626.80/components/prefs/pref_registry_simple.h?autodive=0%2F%2F%2F

Comment (SergeyZhukovsky) [reply vendor/bat-native-ads/src/bat/ads/internal/static_values.h]: sounds good

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 4]
```diff
@@ -692,16 +692,22 @@ void AdsImpl::MaybeClassifyPage(
   MaybeGenerateAdReportingLoadEvent(url, classification);
 }
 
-bool AdsImpl::ShouldClassifyPages() const {
-  auto locale = ads_client_->GetLocale();
-  auto region = helper::Locale::GetRegionCode(locale);
+bool AdsImpl::ShouldClassifyPagesIfTargeted() const {
+  const auto locale = ads_client_->GetLocale();
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: legacy code which will be changed as part of future refactor as not part of this issue. Only added `const` in this PR

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 5]
```diff
@@ -1063,8 +1063,8 @@ void RewardsDOMHandler::GetAdsData(const base::ListValue *args) {
 
   base::DictionaryValue ads_data;
 
-  auto is_supported_region = ads_service_->IsSupportedRegion();
-  ads_data.SetBoolean("adsIsSupported", is_supported_region);
+  auto is_supported_locale = ads_service_->IsSupportedLocale();
```
[the remark this answers is not in the document]
Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Legacy code to be fixed in a future refactor ticket as not related to this change. However new code will/does follow the Google style guide going forward

File: vendor/bat-native-ads/src/bat/ads/ads.cc [hunk 6]
```diff
@@ -21,22 +21,49 @@ const char _catalog_resource_name[] = "catalog.json";
...
-bool Ads::IsSupportedRegion(const std::string& locale) {
-  auto region = GetRegion(locale);
+bool Ads::IsSupportedLocale(
+    const std::string& locale) {
+  const auto region = GetRegion(locale);
 
-  auto it = kSupportedRegions.find(region);
-  if (it == kSupportedRegions.end()) {
-    return false;
+  for (const auto& schema : kSupportedRegionsSchemas) {
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/ads.cc]: We need to iterate all `schema_version` elements in the `std::map`

**Pull request merged**